### PHASE 4 — Thursday | Institutional Partnership Risk

---

`Compliance Update:` For regulatory reporting, default is defined as loans 180+ days past due OR loans with write-off flag. This differs from operational definitions.

`Business Request from Partnership Director:`
"Some of our partner institutions are consistently sending us students who default. I need to know which universities and colleges have the worst track records."


`Thursday Mindset Unlock`: 
Today's output has legal weight. A recommendation to suspend a partnership affects contracts, relationships, and revenue. That means: you cannot recommend a suspension based on insufficient data. An institution with 3 defaults and 10 total loans has a 30% default rate — but that 30% is meaningless. Your job today is to be precise, not dramatic. Volume gates protect the business from acting on noise.

`PARTNERSHIP DIRECTOR'S AUDIT — "Which institutions should we cut?"`

Some partner institutions are consistently sending EduFin borrowers who default. The partnership director needs a clear, legally defensible list of institutions to: suspend, put on review, retain, or reward — based on their actual default performance. Generic rankings are not enough. Each decision must be justified.


`CFO Email - Strategic Clarification`

- From: Rajesh Malhotra, CFO
- To: Data Analytics Team
- Date: 5 hours ago

Quick reminder as you analyze partner performance: our goal is maximizing portfolio profitability, not just minimizing default rates. A partner generating ₹50 Cr in disbursements with 15% defaults (₹7.5 Cr loss) still contributes ₹42.5 Cr net - that's better than a 'safe' partner with ₹10 Cr and 3% defaults (₹9.7 Cr net).

Factor in revenue impact before recommending any partnership suspensions.

Start Spark Session

---

In [1]:
## Import dependencies and create Spark session
import time
from pyspark.sql import SparkSession

spark = SparkSession.builder.getOrCreate()
spark.sparkContext.setLogLevel("ERROR")
spark

Using Spark's default log4j profile: org/apache/spark/log4j2-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
26/05/27 08:06:15 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable


Global Variables

---

In [2]:
N = 25  # Number of rows to show in results
schema = "edufin_national"  # edufin_small or edufin_national
FINAL_DASHBOARD_PATH = "../dashboards"

Load the data

---

In [3]:
## Load the identified datasets and create a temporary views for SQL queries
loans = spark.read.csv(f"../datasets/{schema}/loans.csv", header=True)
institutions = spark.read.csv(f"../datasets/{schema}/institutions.csv", header=True)
customers = spark.read.csv(f"../datasets/{schema}/customers.csv", header=True)

loans.createOrReplaceTempView("loans")
institutions.createOrReplaceTempView("institutions")
customers.createOrReplaceTempView("customers")

print("Loans dataset:")
loans.show(n=5)
print("Customers dataset:")
customers.show(n=5)
print("Institutions dataset:")
institutions.show(n=5)

Loans dataset:
+-------+-----------+--------------+-----------+-------------+-----------------+----------------+-------------+-----------+------------+---------------+------------+
|loan_id|customer_id|institution_id|loan_amount|interest_rate|disbursement_date|application_date|maturity_date|loan_status|default_date|   loan_purpose|program_type|
+-------+-----------+--------------+-----------+-------------+-----------------+----------------+-------------+-----------+------------+---------------+------------+
|      1|      67635|            47|  257163.56|        12.66|       2023-04-19|      2023-03-13|   2026-04-03|     Active|        null|         Hostel|    Domestic|
|      2|      53320|            28|  455070.21|        13.03|       2022-08-31|      2022-08-09|   2025-08-15|     Active|        null|Living Expenses|    Domestic|
|      3|      96263|            43|   230634.4|        14.66|       2023-10-18|      2023-08-31|   2028-09-21|     Active|        null|         Hostel|   

Query 4A (BRD): Institutional Volume Analysis

---

- Calculate: Loan volume and total funding per institution.
- Business Purpose: Identify key partners by size.

In [4]:
query = """
SELECT 
    i.institution_id AS `Institution ID`,
    i.institution_name AS `Institution Name`,
    COUNT(*) AS `Loans`,
    CONCAT(ROUND(SUM(l.loan_amount) / 10000000.0, 2), ' Cr') AS `Loan Amount`
FROM institutions i
INNER JOIN loans l
    ON i.institution_id = l.institution_id
GROUP BY i.institution_id, i.institution_name
ORDER BY `Loans`DESC;
"""
start = time.perf_counter()
spark.sql(query).show(n=N, truncate=False)
print(f"Time: {(time.perf_counter() - start):.3f} seconds")

+--------------+--------------------------------+-----+-----------+
|Institution ID|Institution Name                |Loans|Loan Amount|
+--------------+--------------------------------+-----+-----------+
|13            |State Medical University        |10259|459.15 Cr  |
|4             |Advanced Arts & Science College |10166|451.9 Cr   |
|34            |Modern Management Institute     |10165|461.35 Cr  |
|24            |Modern Law College              |10151|450.19 Cr  |
|3             |State Management University     |10143|458.34 Cr  |
|28            |Regional Medical University     |10134|454.79 Cr  |
|23            |Advanced Technology University  |10121|446.03 Cr  |
|37            |Advanced Arts & Science College |10114|459.27 Cr  |
|26            |State Arts & Science University |10108|456.33 Cr  |
|2             |Central Law University          |10107|452.22 Cr  |
|39            |State Law College               |10106|452.39 Cr  |
|6             |Advanced Commerce Institute     

STEP 4A (Workbook) — Open Metric Design: "Partner Health Score"

---

What you're doing: The Partnership Director wants to avoid "binary" decisions (Suspend/Retain). They want a score. You must define the components of this score using SQL math.

> Define your 'Partner Health Score' Component Logic:

- (1 - default_rate / 100) * volmune_multiplier * 100

-> default_rate is weighted as the primary factor because a partner who defaults often signals a behavioral problem — bad screening, weak student selection, or structural issues that will keep producing defaults regardless of loan size.volume_multiplier is secondary, acting as a reward modifier: a partner with low defaults and high loan volume earns a higher score, reflecting both reliability and scale.

> Your SQL Implementation (One row per institution):

In [5]:
query = """
WITH inst_summary AS (
    SELECT 
        i.institution_id,
        i.institution_name,
        i.institution_type,
        COUNT(*) AS loan_count,
        COUNT(DISTINCT l.customer_id) AS customers,
        SUM(CASE WHEN l.loan_status = 'Defaulted' THEN 1 ELSE 0 END) AS default_count
    FROM institutions i
    INNER JOIN loans l
        ON i.institution_id = l.institution_id
    GROUP BY i.institution_id, i.institution_name, i.institution_type
),
multiplier_data AS (
    SELECT
        *,
        (default_count * 100.0 / loan_count)  AS default_rate,
        CASE
            WHEN loan_count >= 200 THEN 1.5
            WHEN loan_count >= 50 THEN 1.0
            ELSE 0.5
        END AS volume_multiplier
    FROM inst_summary
)
SELECT 
    institution_id AS `Institution Id`,
    institution_name AS `Institution Name`,
    institution_type AS `Institution Type`,
    customers AS `Customers`,
    loan_count AS Loans,
    default_count AS `Default Count`,
    CONCAT(ROUND(default_rate, 2), '%') AS `Default Rate (%)`,
    ROUND((1 - default_rate / 100) * volume_multiplier * 100, 2) AS `Health Score`
FROM multiplier_data
ORDER BY `Health Score` DESC;
"""
start = time.perf_counter()
spark.sql(query).show(n=25, truncate=False)
print(f"Time: {(time.perf_counter() - start):.3f} seconds")

+--------------+---------------------------------+----------------+---------+-----+-------------+----------------+------------+
|Institution Id|Institution Name                 |Institution Type|Customers|Loans|Default Count|Default Rate (%)|Health Score|
+--------------+---------------------------------+----------------+---------+-----+-------------+----------------+------------+
|17            |National Law Institute           |Institute       |9509     |10026|0            |0.00%           |150.00      |
|42            |Advanced Commerce University     |University      |9435     |9910 |1155         |11.65%          |132.52      |
|37            |Advanced Arts & Science College  |College         |9634     |10114|1195         |11.82%          |132.28      |
|35            |Premier Law College              |College         |9436     |9932 |1187         |11.95%          |132.07      |
|1             |National Technology College      |College         |9470     |9962 |1196         |12.01% 

Query 4B (BRD): Partner Performance Metrics

---

- Calculate: Default rate per university/college.
- Business Purpose: Measure quality of students from each partner.
- Filter: Minimum 50 loans to ensure statistical significance.

In [6]:
query = """
SELECT 
    i.institution_type AS `Institution Type`,
    COUNT(*) AS `Loans`,
    SUM(CASE WHEN l.loan_status = 'Defaulted' THEN 1 ELSE 0 END) AS `Default Count`,
    CONCAT(ROUND(SUM(CASE WHEN l.loan_status = 'Defaulted' THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 2), '%') AS `Default Rate (%)`
FROM institutions i
INNER JOIN loans l
    ON i.institution_id = l.institution_id
GROUP BY i.institution_type
HAVING COUNT(*) > 50
ORDER BY SUM(CASE WHEN l.loan_status = 'Defaulted' THEN 1 ELSE 0 END) * 100 / COUNT(*)  DESC;
"""
start = time.perf_counter()
spark.sql(query).show(n=N, truncate=False)
print(f"Time: {(time.perf_counter() - start):.3f} seconds")

+----------------+------+-------------+----------------+
|Institution Type|Loans |Default Count|Default Rate (%)|
+----------------+------+-------------+----------------+
|Institute       |119861|17755        |14.81%          |
|College         |139907|19306        |13.80%          |
|University      |240232|32229        |13.42%          |
+----------------+------+-------------+----------------+

Time: 0.719 seconds


STEP 4B (Workbook) — Default Rate Per Institution

---

What you're doing: Calculate the default rate for each institution. Sort by rate descending. This is the starting point — but remember, the ending point is a multi-criteria decision, not a single ranking.

*Note at least 2 institutions where rate and financial impact tell different stories. This contrast is the core insight of Thursday.*

In [7]:
query = """
WITH default_stats AS (
    SELECT 
        i.institution_id,
        i.institution_name,
        COUNT(*) AS loan_count,
        SUM(CASE WHEN l.loan_status = 'Defaulted' THEN 1 ELSE 0 END) AS default_count,
        ROUND(default_count * 100.0 / loan_count, 2) AS default_rate
    FROM institutions i
    INNER JOIN loans l
        ON i.institution_id = l.institution_id
    GROUP BY i.institution_id, i.institution_name
    HAVING COUNT(l.loan_id) >= 50
)
SELECT
    institution_id AS `Institution ID`,
    institution_name AS `Institution Name`,
    loan_count AS `Loans`,
    default_count AS `Default Count`,
    CONCAT(default_rate, '%') AS `Default Rate (%)`
FROM default_stats
ORDER BY default_rate DESC, default_count DESC;
"""
start = time.perf_counter()
spark.sql(query).show(n=N, truncate=False)
print(f"Time: {(time.perf_counter() - start):.3f} seconds")

+--------------+--------------------------------+-----+-------------+----------------+
|Institution ID|Institution Name                |Loans|Default Count|Default Rate (%)|
+--------------+--------------------------------+-----+-------------+----------------+
|29            |State Commerce Institute        |10005|2088         |20.87%          |
|33            |Central Engineering Institute   |9878 |2050         |20.75%          |
|34            |Modern Management Institute     |10165|2091         |20.57%          |
|31            |Regional Commerce University    |9915 |2039         |20.56%          |
|30            |National Engineering Institute  |9889 |2017         |20.40%          |
|32            |National Commerce Institute     |9947 |2028         |20.39%          |
|50            |Government Medical College      |9780 |1967         |20.11%          |
|28            |Regional Medical University     |10134|1737         |17.14%          |
|25            |Central Commerce College   

Query 4C (BRD): Financial Loss Attribution

---

- Calculate: Total value of defaulted loans attributed to each institution.
- Business Purpose: Determine which partnerships are costing EduFin money.

In [8]:
query = """
SELECT 
    i.institution_id AS `Institution ID`,
    i.institution_name AS `Institution Name`,
    CONCAT(ROUND(SUM(l.loan_amount) / 10000000.0, 2), ' Cr') AS `Loan Amount`,
    CONCAT(ROUND(SUM(CASE WHEN l.loan_status = 'Defaulted' THEN l.loan_amount * 1.0 ELSE 0 END) / 10000000.0, 2), ' Cr') AS `Default Amount`,
    CONCAT(ROUND(SUM(CASE WHEN l.loan_status = 'Defaulted' THEN l.loan_amount * 1.0 ELSE 0 END) * 100 / SUM(l.loan_amount), 2), '%') AS `Loss Rate (%)`
FROM institutions i
INNER JOIN loans l
    ON i.institution_id = l.institution_id
GROUP BY i.institution_id, i.institution_name
ORDER BY SUM(CASE WHEN l.loan_status = 'Defaulted' THEN l.loan_amount * 1.0 ELSE 0 END) DESC;
"""
start = time.perf_counter()
spark.sql(query).show(n=N, truncate=False)
print(f"Time: {(time.perf_counter() - start):.3f} seconds")

+--------------+--------------------------------+-----------+--------------+-------------+
|Institution ID|Institution Name                |Loan Amount|Default Amount|Loss Rate (%)|
+--------------+--------------------------------+-----------+--------------+-------------+
|34            |Modern Management Institute     |461.35 Cr  |88.54 Cr      |19.19%       |
|31            |Regional Commerce University    |448.56 Cr  |85.99 Cr      |19.17%       |
|30            |National Engineering Institute  |451.53 Cr  |85.15 Cr      |18.86%       |
|29            |State Commerce Institute        |446.11 Cr  |83.96 Cr      |18.82%       |
|33            |Central Engineering Institute   |447.13 Cr  |83.5 Cr       |18.68%       |
|32            |National Commerce Institute     |454.88 Cr  |82.81 Cr      |18.2%        |
|50            |Government Medical College      |437.3 Cr   |79.94 Cr      |18.28%       |
|28            |Regional Medical University     |454.79 Cr  |68.86 Cr      |15.14%       |

STEP 4C (Workbook) — Financial Loss Attribution

--- 

What you're doing: Calculate total defaulted loan value per institution in Crores. This converts performance rate into financial cost — the number that justifies partnership decisions to the CFO.

*Which institution has the highest defaulted exposure in Crores? Is it the same institution with the highest default rate?*

In [9]:
query = """
WITH financial_exposure AS (
    SELECT 
        i.institution_id,
        i.institution_name,
        SUM(l.loan_amount) / 10000000.0 AS loan_amount,
        ROUND(SUM(CASE WHEN l.loan_status = 'Overdue' THEN l.loan_amount * 1.0 ELSE 0 END) / 10000000.0, 2) AS overdue_amount,
        ROUND(SUM(CASE WHEN l.loan_status = 'Defaulted' THEN l.loan_amount * 1.0 ELSE 0 END) / 10000000.0, 2) AS default_amount
    FROM institutions i
    INNER JOIN loans l
        ON i.institution_id = l.institution_id
    GROUP BY i.institution_id, i.institution_name
)
SELECT
    institution_id AS `Institution ID`,
    institution_name AS `Institution Name`,
    CONCAT(ROUND(loan_amount, 2), ' Cr') AS `Loan Amount`,
    CONCAT(ROUND(overdue_amount, 2), ' Cr') AS `Overdue Amount`,
    CONCAT(ROUND(default_amount, 2), ' Cr') AS `Default Amount`,
    CONCAT(ROUND(overdue_amount + default_amount, 2), ' Cr') As `At-Risk Exposure`
FROM financial_exposure
ORDER BY default_amount DESC;
"""
start = time.perf_counter()
spark.sql(query).show(n=N, truncate=False)
print(f"Time: {(time.perf_counter() - start):.3f} seconds")

+--------------+--------------------------------+-----------+--------------+--------------+----------------+
|Institution ID|Institution Name                |Loan Amount|Overdue Amount|Default Amount|At-Risk Exposure|
+--------------+--------------------------------+-----------+--------------+--------------+----------------+
|34            |Modern Management Institute     |461.35 Cr  |34.62 Cr      |88.54 Cr      |123.16 Cr       |
|31            |Regional Commerce University    |448.56 Cr  |32.84 Cr      |85.99 Cr      |118.83 Cr       |
|30            |National Engineering Institute  |451.53 Cr  |36.91 Cr      |85.15 Cr      |122.06 Cr       |
|29            |State Commerce Institute        |446.11 Cr  |37.21 Cr      |83.96 Cr      |121.17 Cr       |
|33            |Central Engineering Institute   |447.13 Cr  |34.52 Cr      |83.5 Cr       |118.02 Cr       |
|32            |National Commerce Institute     |454.88 Cr  |36.59 Cr      |82.81 Cr      |119.4 Cr        |
|50            |Gov

Query 4D: Partnership Tier & Risk Classification

---

- Calculate: Categorize partners (Preferred vs. Probation vs. Blacklist).
- Business Purpose: Automated partnership review status.
- Rules: Default rate > 15% = Blacklist Candidate.

In [10]:
query = """
WITH default_stats AS (
    SELECT 
        i.institution_id,
        i.institution_name,
        i.institution_type,
        COUNT(*) AS loan_count,
        SUM(CASE WHEN l.loan_status = 'Defaulted' THEN 1 ELSE 0 END) AS default_count,
        ROUND(default_count * 100.0 / loan_count, 2) AS default_rate
    FROM institutions i
    INNER JOIN loans l  
        ON i.institution_id = l.institution_id
    GROUP BY i.institution_id, i.institution_name, i.institution_type
)
SELECT
    institution_id AS `Institution ID`,
    institution_name AS `Institution Name`, 
    institution_type AS `Institution Type`,
    CONCAT(default_rate, '%') AS `Default Rate (%)`,
    CASE 
        WHEN default_rate > 15 THEN 'Blacklist'
        WHEN default_rate > 12 THEN 'Probation'
        ELSE 'Preferred'
    END AS `Partnership Flag`
FROM default_stats
ORDER BY default_rate DESC;
"""

start = time.perf_counter()
spark.sql(query).show(n=N, truncate=False)
print(f"Time: {(time.perf_counter() - start):.3f} seconds")

+--------------+--------------------------------+----------------+----------------+----------------+
|Institution ID|Institution Name                |Institution Type|Default Rate (%)|Partnership Flag|
+--------------+--------------------------------+----------------+----------------+----------------+
|29            |State Commerce Institute        |Institute       |20.87%          |Blacklist       |
|33            |Central Engineering Institute   |Institute       |20.75%          |Blacklist       |
|34            |Modern Management Institute     |Institute       |20.57%          |Blacklist       |
|31            |Regional Commerce University    |University      |20.56%          |Blacklist       |
|30            |National Engineering Institute  |Institute       |20.40%          |Blacklist       |
|32            |National Commerce Institute     |Institute       |20.39%          |Blacklist       |
|50            |Government Medical College      |College         |20.11%          |Blacklis

STEP 4D (Workbook) — Minimum Data Threshold Gate

---

What you're doing: Identify institutions with insufficient data — fewer than 50 loans. These must be explicitly excluded from the suspension recommendation and flagged as "Insufficient Data." This is statistical discipline, not leniency.

*Write down: how many institutions are excluded by the 50-loan threshold? If you lowered it to 20 — would any new institutions enter the suspension candidate list?*

In [11]:
query = """
WITH default_stats AS (
    SELECT 
        i.institution_id,
        i.institution_name,
        i.institution_type,
        COUNT(*) AS loan_count,
        SUM(CASE WHEN l.loan_status = 'Defaulted' THEN 1 ELSE 0 END) AS default_count,
        ROUND(default_count * 100.0 / loan_count, 2) AS default_rate
    FROM institutions i
    INNER JOIN loans l  
        ON i.institution_id = l.institution_id
    GROUP BY i.institution_id, i.institution_name, i.institution_type
    HAVING COUNT(l.loan_id) >= 50
)
SELECT
    institution_id AS `Institution ID`,
    institution_name AS `Institution Name`, 
    institution_type AS `Institution Type`,
    CONCAT(default_rate, '%') AS `Default Rate (%)`,
    CASE 
        WHEN default_rate > 15 THEN 'Blacklist'
        WHEN default_rate > 12 THEN 'Monitor'
        ELSE 'Expand'
    END AS `Risk Classification`
FROM default_stats
ORDER BY default_rate DESC;
"""

start = time.perf_counter()
spark.sql(query).show(n=N, truncate=False)
print(f"Time: {(time.perf_counter() - start):.3f} seconds")

+--------------+--------------------------------+----------------+----------------+-------------------+
|Institution ID|Institution Name                |Institution Type|Default Rate (%)|Risk Classification|
+--------------+--------------------------------+----------------+----------------+-------------------+
|29            |State Commerce Institute        |Institute       |20.87%          |Blacklist          |
|33            |Central Engineering Institute   |Institute       |20.75%          |Blacklist          |
|34            |Modern Management Institute     |Institute       |20.57%          |Blacklist          |
|31            |Regional Commerce University    |University      |20.56%          |Blacklist          |
|30            |National Engineering Institute  |Institute       |20.40%          |Blacklist          |
|32            |National Commerce Institute     |Institute       |20.39%          |Blacklist          |
|50            |Government Medical College      |College        

Query 4E (BRD): Partnership Review Dashboard

---

- Calculate: Comprehensive partner scorecard.
- Business Purpose: Support contract renegotiation or termination.
- Sorting: Highest Default Rate first.
- REQUIRED Text Summary: List of top 5 institutions to blacklist immediately.

In [12]:
query = """
WITH default_stats AS (
    SELECT 
        i.institution_id,
        i.institution_name,
        i.institution_type,
        COUNT(*) AS loan_count,
        SUM(CASE WHEN l.loan_status = 'Defaulted' THEN 1 ELSE 0 END) AS default_count
    FROM institutions i
    INNER JOIN loans l  
        ON i.institution_id = l.institution_id
    GROUP BY i.institution_id, i.institution_name, i.institution_type
    HAVING COUNT(l.loan_id) >= 50
),
multiplier_data AS (
    SELECT
        *,
        (default_count * 100.0 / loan_count)  AS default_rate,
        CASE
            WHEN loan_count >= 200 THEN 1.5
            WHEN loan_count >= 50 THEN 1.0
            ELSE 0.5
        END AS volume_multiplier
    FROM default_stats
)
SELECT
    institution_id AS `Institution ID`,
    institution_name AS `Institution Name`, 
    institution_type AS `Institution Type`,
    CONCAT(ROUND(default_rate, 2), '%') AS `Default Rate (%)`,
    ROUND((1 - default_rate / 100) * volume_multiplier * 100, 2) AS `Health Score`,
    CASE 
        WHEN default_rate > 15 THEN 'Blacklist'
        WHEN default_rate > 8 THEN 'Monitor'
        ELSE 'Expand'
    END AS `Partner Flag`
FROM multiplier_data
ORDER BY default_rate DESC;
"""
start = time.perf_counter()
spark.sql(query).show(n=N, truncate=False)
print(f"Time: {(time.perf_counter() - start):.3f} seconds")

+--------------+--------------------------------+----------------+----------------+------------+------------+
|Institution ID|Institution Name                |Institution Type|Default Rate (%)|Health Score|Partner Flag|
+--------------+--------------------------------+----------------+----------------+------------+------------+
|29            |State Commerce Institute        |Institute       |20.87%          |118.70      |Blacklist   |
|33            |Central Engineering Institute   |Institute       |20.75%          |118.87      |Blacklist   |
|34            |Modern Management Institute     |Institute       |20.57%          |119.14      |Blacklist   |
|31            |Regional Commerce University    |University      |20.56%          |119.15      |Blacklist   |
|30            |National Engineering Institute  |Institute       |20.40%          |119.41      |Blacklist   |
|32            |National Commerce Institute     |Institute       |20.39%          |119.42      |Blacklist   |
|50       

STEP 4E (Workbook) — Partnership Review Dashboard (Final Output)

---

What you're doing: Combine all metrics into one output per institution: total loans, default rate, defaulted exposure, partner_flag (Suspend/Review/Retain/Reward), and minimum_threshold_flag. This is the final document the Partnership Director takes to the contract renegotiation meeting.

*Read your top suspended institution's row aloud. Would a legal team accept this as grounds for contract suspension? Does it include: institution ID, rate, volume, exposure, and threshold exceeded?*

In [13]:
dashboard_query = """
WITH institution_summary AS (
    SELECT 
        i.institution_id,
        i.institution_name,
        i.institution_type,
        COUNT(*) AS loan_count,
        COUNT(DISTINCT l.customer_id) AS customers,
        SUM(CASE WHEN l.loan_status = 'Defaulted' THEN 1 ELSE 0 END) AS default_count,
        ROUND(SUM(CASE WHEN l.loan_status = 'Defaulted' THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 2) AS default_rate,
        ROUND(SUM(CASE WHEN l.loan_status = 'Defaulted' THEN l.loan_amount * 1.0 ELSE 0 END) / 10000000.0, 2) AS default_amount
    FROM institutions i
    INNER JOIN loans l
        ON i.institution_id = l.institution_id
    GROUP BY i.institution_id, i.institution_name, i.institution_type
    HAVING COUNT(l.loan_id) >= 50
),
multiplier_data AS (
    SELECT 
        *,
        CASE
            WHEN loan_count >= 200 THEN 1.5
            WHEN loan_count >= 50 THEN 1.0
            ELSE 0.5
        END AS volume_multiplier
    FROM institution_summary
)
SELECT
    institution_id AS `Institution ID`,
    institution_name AS `Institution Name`,
    institution_type AS `Institution Type`,
    loan_count AS `Loans`,
    customers AS `Customers`,
    CONCAT(default_amount, ' Cr') AS `Default Amount`,
    CONCAT(default_rate, '%') AS `Default Rate (%)`,
    ROUND((1 - default_rate / 100) * volume_multiplier * 100, 2) AS `Health Score`,
    CASE 
        WHEN default_rate > 15 THEN 'Blacklist'
        WHEN default_rate > 8 THEN 'Monitor'
        ELSE 'Expand'
    END AS `Partner Flag`
FROM multiplier_data
ORDER BY default_rate DESC;
"""
start = time.perf_counter()
spark.sql(dashboard_query).show(n=N, truncate=False)
print(f"Time: {(time.perf_counter() - start):.3f} seconds")

+--------------+--------------------------------+----------------+-----+---------+--------------+----------------+------------+------------+
|Institution ID|Institution Name                |Institution Type|Loans|Customers|Default Amount|Default Rate (%)|Health Score|Partner Flag|
+--------------+--------------------------------+----------------+-----+---------+--------------+----------------+------------+------------+
|29            |State Commerce Institute        |Institute       |10005|9501     |83.96 Cr      |20.87%          |118.70      |Blacklist   |
|33            |Central Engineering Institute   |Institute       |9878 |9420     |83.5 Cr       |20.75%          |118.88      |Blacklist   |
|34            |Modern Management Institute     |Institute       |10165|9660     |88.54 Cr      |20.57%          |119.15      |Blacklist   |
|31            |Regional Commerce University    |University      |9915 |9433     |85.99 Cr      |20.56%          |119.16      |Blacklist   |
|30          

Save the Final Dashboard

---

In [14]:
import os
import shutil


def save_dashboard(output_path, dashboard_name, dashboard):
    final_output_path = output_path + "/" + dashboard_name
    if os.path.exists(final_output_path):
        print(f"⚠️  File already exists at: {final_output_path}")
    else:
        # Create a temporary directory for Spark output
        temp_dir = final_output_path + "_temp"
        # Write to temporary directory
        dashboard.coalesce(1).write.mode("overwrite").option("header", "true").csv(
            temp_dir
        )
        # Find and move the actual CSV file
        part_files = [f for f in os.listdir(temp_dir) if f.startswith("part-")]
        if part_files:
            # Move the part file to the final location with .csv extension
            shutil.move(os.path.join(temp_dir, part_files[0]), final_output_path)
            shutil.rmtree(temp_dir)  # Delete temporary directory
            print(f"✓ CSV saved to: {final_output_path}")
        else:
            print("❌ Error: No CSV file generated")

In [15]:
dashboard = spark.sql(dashboard_query)
save_dashboard(
    output_path=FINAL_DASHBOARD_PATH,
    dashboard_name="phase4_dashboard.csv",
    dashboard=dashboard,
)

⚠️  File already exists at: ../dashboards/phase4_dashboard.csv


BUILD YOUR CUSTOM COLUMNS — Required for Thursday Submission

---


> NEW TECHNIQUE — Correlated Subqueries: Phase 1 = CASE. Phase 2 = Window Functions. Phase 3 = CTEs. Today you unlock correlated subqueries — a subquery that references a column from the outer query. This lets you compare each institution against the portfolio-wide baseline row by row.

- Column 1 — partner_flag: Compare each institution's default rate against the portfolio-wide average default rate using a correlated subquery. Suspend where default_rate > 1.5× portfolio average AND total_loans >= 50. Review where > 1.0× average. Retain where below average. Reward where < 0.5× average AND total_loans >= 50.

- Column 2 — anomaly_flag: Use EXISTS to flag institutions where at least one borrower has CIBIL > 900 (data anomaly). This is cross-table validation. Mark as: "Data Quality Review Required" or "Data Clean".

Your partner_flag with correlated subquery (must reference portfolio average dynamically):

In [16]:
query = """
CASE 
    WHEN default_rate >= 1.5 * (SELECT AVG(default_rate) FROM institution_summary) THEN 'Suspend'
    WHEN default_rate >= (SELECT AVG(default_rate) FROM institution_summary) THEN 'Review'
    WHEN default_rate >= 0.5 * (SELECT AVG(default_rate) FROM institution_summary) THEN 'Reward'
    ELSE 'Retain'
END AS partner_flag
"""

Your anomaly_flag using EXISTS (cross-table validation):

In [17]:
query = """
CASE WHEN EXISTS (
    SELECT 1 FROM customers c
    JOIN loans l
        ON c.customer_id = l.customer_id
    WHERE l.institution_id = institution_summary.institution_id AND c.cibil_score > 900
)
THEN 'Data Quality Review Required' 
ELSE 'Data Clean' END AS anomaly_flag
"""

Why is comparing against the portfolio average better than a fixed threshold? (2–3 sentences):

-> A fixed threshold like default_rate > 15% assumes you already know what "bad" looks like before seeing the data, but that number is arbitrary and breaks down across different economic conditions, loan types, or time periods. Comparing against the portfolio average is self-calibrating: if the whole portfolio is performing poorly (say average is 18%), a 15% institution is actually your best partner, not a suspension candidate. It also means your flags automatically adapt as the portfolio changes — no one needs to manually update hardcoded thresholds every quarter.

### BEFORE YOU SUBMIT — Answer These Three Questions

---

1. The institution with 0 defaults — what action did you assign it, and why? Write your classification AND your reasoning. ("I assigned it X because Y, not Z because A.")

-> 'National Law Institute' (Institution ID 17) is the only institution with 0 defaults. I classified the institution as 'Reward' and not 'Retain', 'Review' OR 'Suspend', because its default rate of 0.00% falls below 0.5x the portfolio average, making it the strongest performer/institution in the entire dataset. I did not classify it as 'Retain' because Retain is reserved for institutions between 0.5x and 1x the average, which still carry moderate risk. 'Review' and 'Suspend' are ruled out entirely since those require a default rate at or above the portfolio average.

---

2. Write your Suspension Justification for the top-ranked institution to be suspended. Format it as if going to a legal review: Institution name, default rate, loan count, total defaulted exposure, which threshold was exceeded, recommended action, and what monitoring must happen post-suspension.

- Institution: State Commerce Institute (Institution ID: 29)
- Default Rate: 20.87%
- Total Loans Disbursed: 10005
- Total Defaulted Exposure: 83.96 Crores
- Threshold Exceeded: 20.87% exceeds the 1.5x portfolio average threshold
- Recommended Action:  Immediate suspension of new loan disbursements pending full portfolio audit. No new student loan applications to be processed through this institution until review is complete.
- Post-Suspension Monitoring Requirements:
    - 83.96 Cr in defaulted exposure must be assigned to collections with a 90-day recovery report due to the Partnership Director
    - Institution returned Data Clean on the anomaly check, meaning no CIBIL >900 irregularities were detected; however a manual data audit is still recommended given the volume
    - Partnership agreement to be placed under legal hold; no contract renewals or amendments permitted during suspension period.

---

3. If you were the Partnership Director and received your final output — what is the FIRST call you make? Who do you call, about which institution, and what specifically do you say in that call?

-> If i were the partnership director that receive my own final output, then i would first try to call the institution who had 0% default rate despite having 10026 loans. This result did not came from an rouding error, but due to the structural difference of how you guys are operating compared to the other institutions. Before I spend the next few weeks on suspension hearings for the bottom of this portfolio, I want to understand what you're doing differently like, what's your screening process, your student selection criteria, your collections touchpoints, whatever it is. Because if we can document it and replicate it, that's worth more to this organization than suspending 10 bad partners.

---